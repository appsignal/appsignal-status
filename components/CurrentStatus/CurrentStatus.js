import PropTypes from "prop-types"
import Link from "next/link"

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faCheck, faExclamationTriangle } from "@fortawesome/free-solid-svg-icons"

const iconMapping = {
  "UP": {
    outerStyle: "bg-green-200",
    innerStyle: "bg-green-500",
    icon: faCheck
  },
  "DOWN": {
    outerStyle: "bg-red-200",
    innerStyle: "bg-red-500",
    icon: faExclamationTriangle
  },
}

const CurrentStatus = ({ status }) => {
  const renderIcon = () => {
    const outerStyle = iconMapping[status].outerStyle
    const innerStyle = iconMapping[status].innerStyle
    const icon = iconMapping[status].icon

    return (
      <div className={`flex items-center justify-center h-16 sm:h-20 w-16 sm:w-20 mx-auto rounded-full ${outerStyle}`}>
        <div className={`flex items-center justify-center h-12 sm:h-16 w-12 sm:w-16 rounded-full ${innerStyle}`}>
          <FontAwesomeIcon className="text-white text-2xl sm:text-3xl" icon={icon} fixedWidth />
        </div>
      </div>
    )
  }

  return (
    <section className="mt-16">
      <div className="container text-center">
        { renderIcon() }
        <h1 className="mt-4 mb-3 c_h-heading c_h-heading--3xl sm:c_h-heading--4xl">
          { status == "UP" ? "No known issues" : "Weâ€™re experiencing some issues"}
        </h1>
        <p className="text-gray-700">
          { status == "UP"
            ? <>Don{"'"}t agree with this? Please <a href="mailto:contact@appsignal.com">let us know</a>.</>
            : <>Last updated at 15th Aug 07:53 CET. Read <Link href="#status-updates"><a>latest status update</a></Link>.</>
          }
        </p>
      </div>
    </section>
  )
}

CurrentStatus.propTypes = {
  status: PropTypes.oneOf(["UP", "DOWN"]).isRequired,
}

export default CurrentStatus
