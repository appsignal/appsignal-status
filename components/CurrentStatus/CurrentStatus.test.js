import { render, screen } from "@testing-library/react";

import statusPageMock from "../../mocks/status_pages/appsignal.json";

import CurrentStatus from "./CurrentStatus";

const build = (props = {}) => {
  return render(<CurrentStatus statusPage={statusPageMock} {...props} />);
};

describe("CurrentStatus", () => {
  test("renders a checkmark with all good text when up", () => {
    build();

    expect(screen.getByRole("img", { hidden: true }).classList).toContain(
      "fa-check"
    );
    expect(screen.getByText("No known issues")).toBeInTheDocument();
  });

  test("renders a down icon and message when down", () => {
    build({
      statusPage: {
        state: "down",
        updates: [
          {
            title: "All ok",
            description: null,
            time: new Date("2021-09-01 08:00"),
          },
          {
            title: "Problems with the paper stock",
            description: "We are printing new paper",
            time: new Date("2021-09-01 12:00"),
          },
        ],
      },
    });

    expect(screen.getByRole("img", { hidden: true }).classList).toContain(
      "fa-exclamation-triangle"
    );

    expect(
      screen.getByText("Problems with the paper stock")
    ).toBeInTheDocument();
    expect(screen.getByText("We are printing new paper")).toBeInTheDocument();
  });

  test("it renders markdown", () => {
    const { container } = build({
      statusPage: {
        state: "down",
        updates: [
          {
            title: "Problems with the paper stock",
            description:
              "# This is heading This is a [link](https://example.com). ",
            time: new Date("2021-09-01 12:00"),
          },
        ],
      },
    });
    expect(container).toMatchSnapshot();

    const status = screen.getAllByTestId("CurrentStatusMarkdownDescription");

    expect(status[0].innerHTML).toContain("<h1>");
    expect(status[0].innerHTML).toContain("<a href");
  });
});
