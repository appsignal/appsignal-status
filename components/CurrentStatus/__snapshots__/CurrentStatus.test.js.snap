// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentStatus it renders markdown 1`] = `
<div>
  <section
    class="mt-16"
  >
    <div
      class="container"
    >
      <div
        class="flex items-center justify-center h-16 sm:h-20 w-16 sm:w-20 mx-auto rounded-full bg-red-200"
      >
        <div
          class="flex items-center justify-center h-12 sm:h-16 w-12 sm:w-16 rounded-full bg-red-500"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-exclamation-triangle fa-w-18 fa-fw text-white text-2xl sm:text-3xl"
            data-icon="exclamation-triangle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <h1
        class="mt-4 mb-3 c_h-heading c_h-heading--3xl sm:c_h-heading--4xl text-center"
      >
        Problems with the paper stock
      </h1>
      <div
        class="text-gray-700 bg-white shadow-sm rounded py-5 px-6"
        data-testid="CurrentStatusMarkdownDescription"
      >
        <h1>
          This is heading This is a 
          <a
            href="https://example.com"
          >
            link
          </a>
          .
        </h1>
      </div>
    </div>
  </section>
</div>
`;
